### PR TITLE
Implement multibox loss wrapper for SSD in Python API.

### PR DESCRIPTION
**Pull request by qingqing01**
Fix https://github.com/PaddlePaddle/Paddle/issues/8252

### CODE REVIEW

File: python/paddle/v2/fluid/layers/detection.py [hunk 1]
```diff
@@ -326,3 +335,328 @@ def _reshape_with_axis_(input, axis=1):
...
+            are called LoD. Then ColToRowMatchDist[i][j] = DistMat[d+LoD[i]][j].
+    """
+    helper = LayerHelper('bipartite_match', **locals())
+    dtype = helper.input_dtype()
+
+    match_indices = helper.create_tmp_variable(dtype='int32')
+    match_distance = helper.create_tmp_variable(dtype=dtype)
+    helper.append_op(
+        type='bipartite_match',
+        inputs={'DistMat': similarity_matrix},
```
Comment (wanghaox) [inline python/paddle/v2/fluid/layers/detection.py]: 为什么inputs不直接叫DistMat，而改为similarity_matrix ？

Comment (wanghaox) [inline python/paddle/v2/fluid/layers/detection.py]: 为什么inputs不直接叫DistMat，而改为similarity_matrix ？

File: python/paddle/v2/fluid/layers/detection.py [hunk 2]
```diff
@@ -326,3 +335,328 @@ def _reshape_with_axis_(input, axis=1):
...
+
+    # 1. Find matched boundding box by prior box.
+    #   1.1 Compute IOU similarity between ground-truth boxes and prior boxes.
+    iou = iou_similarity(x=gt_box, y=prior_box)
+    #   1.2 Compute matched boundding box by bipartite matching algorithm.
+    matched_indices, matched_dist = bipartite_match(iou)
+
+    # 2. Compute confidence for mining hard examples
+    # 2.1. Get the target label based on matched indices
+    gt_label = ops.reshape(x=gt_label, shape=gt_label.shape + (1, ))
```
Comment (wanghaox) [inline python/paddle/v2/fluid/layers/detection.py]: 这里为什么要reshape，而不是让传入的gt_label满足参数要求

File: python/paddle/v2/fluid/layers/detection.py (hunk 1 above)
Comment (qingqing01) [reply python/paddle/v2/fluid/layers/detection.py]: Done,  change to `dist_matrix`.

Comment (qingqing01) [reply python/paddle/v2/fluid/layers/detection.py]: Done,  change to `dist_matrix`.

File: python/paddle/v2/fluid/layers/detection.py [hunk 3]
```diff
@@ -326,3 +335,328 @@ def _reshape_with_axis_(input, axis=1):
...
+            gt_box = layers.data(
+                name='gt_box', shape=[4], lod_level=1, dtype='float32')
+            gt_label = layers.data(
+                name='gt_label', shape=[1], lod_level=1, dtype='float32')
+            loss = layers.ssd_loss(loc, scores, gt_box, gt_label, pb, pbv)
+    """
+
+    helper = LayerHelper('ssd_loss', **locals())
+    dtype = helper.input_dtype()
+    if mining_type == 'hard_example':
```
[the remark this answers is not in the document]
Comment (qingqing01) [reply python/paddle/v2/fluid/layers/detection.py]: Done.

File: python/paddle/v2/fluid/layers/detection.py (hunk 2 above)
Comment (qingqing01) [reply python/paddle/v2/fluid/layers/detection.py]: 1. 用户传入的gt_label是： [Ng, 1]，对于用户是合理的。
2. target_assign op为了是的confidence、localization的assigning统一，check输入的shape必需是3D，因此这里reshape了。
